lib/cretonne/src/licm.rs

```diff
@@ -79,6 +89,32 @@ fn create_pre_header(header: Ebb,
     pre_header
 }
 
+// Detects if a loop header has a natural pre-header.
+//
+// A loop header has a pre-header if there is only one predecessor that the header doesn't
+// dominate.
+fn has_pre_header(layout: &Layout,
+                  cfg: &ControlFlowGraph,
+                  domtree: &DominatorTree,
+                  header: Ebb)
+                  -> Option<Ebb> {
+    let mut result = None;
+    let mut found = false;
+    for &(pred_ebb, last_inst) in cfg.get_predecessors(header) {
```

Comment:
You should verify that `last_inst` really is the terminator of `pred_ebb`. Otherwise you could be inserting code in a place that isn't post-dominated by the loop header.

It's probably also a good idea to add a test case for this special case.